[mlir][Interfaces] `DestinationStyleOpInterface`: Rename `hasTensor/BufferSemantics`

matthias-springer opened the pull request:
Rename interface functions as follows:
* `hasTensorSemantics` -> `hasPureTensorSemantics`
* `hasBufferSemantics` -> `hasPureBufferSemantics`

These two functions return "true" if the op has tensor/buffer operands but not buffer/tensor operands.

Also drop the "ranked" part from the interface, i.e., do not distinguish between ranked/unranked types.

The new function names describe the functions more accurately. They also align their semantics with the notion of "tensor semantics" with the bufferization framework. (An op is supposed to be bufferized if it has tensor operands, and we don't care if it also has memref operands.)

This change is in preparation of #75273, which adds `BufferizableOpInterface::hasTensorSemantics`. By renaming the functions in the `DestinationStyleOpInterface`, we can avoid name clashes between the two interfaces.

jpienaar commented on mlir/include/mlir/Interfaces/DestinationStyleOpInterface.td:
```diff
@@ -17,24 +17,26 @@ def DestinationStyleOpInterface : OpInterface<"DestinationStyleOpInterface"> {
+    Note: This implies that a destination style op without any tensor inits must
+    not have any OpResults.
+
+    An op has "tensor semantics" if it has at least one tensor operand.
```
Should this be a bulleted list?

matthias-springer replied:
I removed the "tensor semantics" and "memref semantics" functions. There are only two "bullets" left now, I think we can leave it as a paragraph.

jpienaar reviewed (commented):
This looks good to me, BUT I don't have as much context as Hanhan or the like :)

MaheshRavishankar reviewed (changes requested):
Is there a use for having interface methods that say if it has any tensor operands or any buffer operands. I dont see any use cases here. Maybe we just drop those methods and make this PR a pure rename.

matthias-springer committed:
[mlir][Interface] `DestinationStyleOpInterface`: Rename `hasTensor/BufferSemantics`

Rename interface functions as follows:
* `hasTensorSemantics` -> `hasPureTensorSemantics`
* `hasBufferSemantics` -> `hasPureBufferSemantics`

These two functions return "true" if the op has tensor/buffer operands but not buffer/tensor operands.

Add two new interface functions:
* `hasTensorSemantics`: Return "true" if the op has tensor operands. Whether the op has buffer operands or not does not matter.
* `hasBufferSemantics`: Return "true" if the op has buffer operands. Whether the op has tensor operands or not does not matter.

Also drop the "ranked" part from the interface, i.e., do not distinguish between ranked/unranked types.

This change aligns the meaning of "tensor semantics" with the bufferization framework. (An op is supposed to be bufferized if it has tensor operands, and we don't care if it also has memref operands.) This change is in preparation of #75273, which adds `BufferizableOpInterface::hasTensorSemantics`.

matthias-springer commented:
Addressed comments: Removed `hasTensorSemantics` and `hasBufferSemantics` functions, which are not used anywhere.

MaheshRavishankar reviewed (approved):
Thanks!
